File: game/models/game.go
package models

import (
	"encoding/json"
	"errors"

	"slices"

	"github.com/google/uuid"
)

type GameStatus string

const (
	GameStatusWaiting  GameStatus = "waiting"
	GameStatusStarted  GameStatus = "started"
	GameStatusFinished GameStatus = "finished"
)

type GamePlayerStatus string

const (
	GamePlayerStatusWaiting  GamePlayerStatus = "waiting"
	GamePlayerStatusActive   GamePlayerStatus = "active"
	GamePlayerStatusInactive GamePlayerStatus = "inactive"
)

type GameError error

var (
	ErrorGameFull            GameError = errors.New("game_full")
	ErrorGamePlayerAlreadyIn GameError = errors.New("game_player_already_in")
	ErrorGamePositionTaken   GameError = errors.New("game_position_taken")
	ErrorGamePlayerNotFound  GameError = errors.New("game_player_not_found")
	ErrorGameNotReady        GameError = errors.New("game_not_ready")
)

type GameType string

const (
	GameTypeHoldem GameType = "holdem"
)

type IPlayable interface {
	Start() error
	End() error
	ProcessAction(action json.RawMessage) error
	CanStart() bool
	DealCards() error
	EvaluateHands() error
	GetGameState() interface{}
}

type GameAction struct {
	PlayerID string          `json:"playerId"`
	Action   string          `json:"action"` // fold, check, call, raise
	Data     json.RawMessage `json:"data"`
}

type Card struct {
	Suit   string `json:"suit"`
	Value  string `json:"value"`
	Hidden bool   `json:"hidden"`
}

type GamePlayer struct {
	Position   int              `json:"position"`
	Balance    int              `json:"balance"`
	LastAction string           `json:"lastAction"`
	Player     Player           `json:"player"`
	Hand       []Card           `json:"hand"`
	Status     GamePlayerStatus `json:"status"`
}

type Game struct {
	ID         string          `json:"id"`
	Status     GameStatus      `json:"status"`
	Players    []*GamePlayer   `json:"players"`
	Playable   IPlayable       `json:"playable"`
	MinBet     int             `json:"minBet"`
	MaxPlayers int             `json:"maxPlayers"`
	ActionChan chan GameAction `json:"-"`
	GameType   GameType        `json:"gameType"`
}

type GameState struct {
	Status  GameStatus    `json:"status"`
	Players []*GamePlayer `json:"players"`
	State   interface{}   `json:"state"`
}

func NewGame(actionChan chan GameAction, maxPlayers int, minBet int, gameType GameType) *Game {
	return &Game{
		ID:         uuid.New().String(),
		Status:     GameStatusWaiting,
		Players:    make([]*GamePlayer, 0),
		MaxPlayers: maxPlayers,
		MinBet:     minBet,
		ActionChan: actionChan,
		GameType:   gameType,
	}
}

func (g *Game) AddPlayer(player *GamePlayer) error {
	if len(g.Players) >= g.MaxPlayers {
		return ErrorGameFull
	}

	for _, p := range g.Players {
		if p.Player.ID == player.Player.ID {
			return ErrorGamePlayerAlreadyIn
		}

		if p.Position == player.Position {
			return ErrorGamePositionTaken
		}
	}

	player.Status = GamePlayerStatusWaiting
	g.Players = append(g.Players, player)
	return nil
}

func (g *Game) RemovePlayer(playerID string) error {
	for i, p := range g.Players {
		if p.Player.ID == playerID {
			g.Players = slices.Delete(g.Players, i, i+1)
			return nil
		}
	}
	return ErrorGamePlayerNotFound
}

func (g *Game) CanStart() bool {
	return len(g.Players) >= 2 && g.Status == GameStatusWaiting
}

func (g *Game) Start() error {
	if !g.CanStart() {
		return ErrorGameNotReady
	}

	err := g.Playable.Start()
	if err != nil {
		return err
	}

	g.Status = GameStatusStarted
	return nil
}

func (g *Game) End() error {
	err := g.Playable.End()
	if err != nil {
		return err
	}

	g.Status = GameStatusFinished
	return nil
}

func (g *Game) GetGameState() GameState {
	return GameState{
		Status:  g.Status,
		Players: g.Players,
		State:   g.Playable.GetGameState(),
	}
}

File: Bot/Models/Game.cs
using Newtonsoft.Json;

namespace Bot.Models;

public class Card
{
    [JsonProperty("suit")] public string Suit { get; set; }
    [JsonProperty("value")] public string Value { get; set; }
    [JsonProperty("hidden")] public bool Hidden { get; set; }
}

public class GamePlayer
{
    [JsonProperty("id")] public string Id { get; set; }
    [JsonProperty("name")] public string Name { get; set; }
    [JsonProperty("cards")] public List<Card> Cards { get; set; } = new();
    [JsonProperty("chips")] public int Chips { get; set; }
    [JsonProperty("bet")] public int Bet { get; set; }
    [JsonProperty("position")] public int Position { get; set; }
    [JsonProperty("active")] public bool Active { get; set; }
    [JsonProperty("folded")] public bool Folded { get; set; }
    [JsonProperty("lastAction")] public string LastAction { get; set; }
}

public class Game
{
    [JsonProperty("id")] public string Id { get; set; }
    [JsonProperty("status")] public string Status { get; set; }
    [JsonProperty("players")] public List<GamePlayer> Players { get; set; } = new();
    [JsonProperty("communityCards")] public List<Card> CommunityCards { get; set; } = new();
    [JsonProperty("pot")] public int Pot { get; set; }
    [JsonProperty("currentBet")] public int CurrentBet { get; set; }
    [JsonProperty("dealerPosition")] public int DealerPosition { get; set; }
    [JsonProperty("currentTurn")] public int CurrentTurn { get; set; }
    [JsonProperty("minBet")] public int MinBet { get; set; }
    [JsonProperty("maxPlayers")] public int MaxPlayers { get; set; }
    [JsonProperty("createdAt")] public DateTime CreatedAt { get; set; }
    [JsonProperty("updatedAt")] public DateTime UpdatedAt { get; set; }
}

public class Room
{
    [JsonProperty("id")] public string Id { get; set; }
    [JsonProperty("name")] public string Name { get; set; }
    [JsonProperty("status")] public string Status { get; set; }
    [JsonProperty("game")] public Game Game { get; set; }
    [JsonProperty("maxPlayers")] public int MaxPlayers { get; set; }
    [JsonProperty("minBet")] public int MinBet { get; set; }
    [JsonProperty("createdAt")] public DateTime CreatedAt { get; set; }
    [JsonProperty("updatedAt")] public DateTime UpdatedAt { get; set; }
}